### PR TITLE
Issue #6693: Make link_field more fully available to custom forms.

### LINKED ISSUE #6693: Make `link_field` more fully available to custom forms
## Description of the bug

Link field form elements are not fully available to custom forms. They can be added to a custom form but result in a variety of warnings:

```
    Warning: Trying to access array offset on value of type null in field_widget_instance() (line 539 of /var/www/html/core/modules/field/field.form.inc).
    Warning: Trying to access array offset on value of type null in link_field_process() (line 853 of /var/www/html/core/modules/link/link.module).
    Warning: Trying to access array offset on value of type null in link_field_process() (line 861 of /var/www/html/core/modules/link/link.module).
    Warning: Trying to access array offset on value of type null in link_field_process() (line 864 of /var/www/html/core/modules/link/link.module).
    Warning: Trying to access array offset on value of type null in link_field_process() (line 867 of /var/www/html/core/modules/link/link.module).
    Warning: Trying to access array offset on value of type null in link_field_process() (line 867 of /var/www/html/core/modules/link/link.module).
    Warning: Trying to access array offset on value of type null in link_field_process() (line 881 of /var/www/html/core/modules/link/link.module).
    Warning: Trying to access array offset on value of type null in link_field_process() (line 883 of /var/www/html/core/modules/link/link.module).
    Warning: Trying to access array offset on value of type null in link_field_process() (line 889 of /var/www/html/core/modules/link/link.module).
```

## Steps To Reproduce

To reproduce the behavior:
1. Create a simple custom form with a `link_field` element.
2. Submit the form

## Actual behavior

The warnings above show up.

## Expected behavior

No warnings.

## Additional information

@docwilmot has identified two commits in the Drupal 7 version of Link which address this issue. I'll submit a PR that includes these, as well as a variety of other `link_field` instance settings.
 - [D7-2718563](https://www.drupal.org/project/link/issues/2718563) by makbul_khan8, jgalletta, idebr, malcomio, joekers, b.lammers, pifagor: Use link field widget in custom form.
 - [D7-3276037](https://www.drupal.org/project/link/issues/3276037) by xlin1003, DamienMcKenna: PHP 7.4  Notice: Trying to access array offset on value of type null in field_widget_instance() when use in custom form.

### PR DESCRIPTION
**Pull request by laryn**
Fixes https://github.com/backdrop/backdrop-issues/issues/6693

### CODE REVIEW

File: core/modules/link/link.module [hunk 1]
```diff
@@ -813,30 +776,84 @@ function link_theme() {
...
+ *       'class' => 'additional-css-class',
+ *       'target' => '_blank',
+ *     ),
+ *   ),
+ * @endcode
  */
 function link_element_info() {
   $elements = array();
   // We use "link_field" here to avoid conflicts with theme_link().
+  // Many of these property are related to the Field API Link field settings.
```
Comment (quicksketch) [inline core/modules/link/link.module]:
```suggestion
  // Many of these properties are related to the Field API Link field settings.
```

File: core/modules/link/tests/link.custom.test [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    $this->backdropPost(NULL, array('link_test_validate_url[url]' => ''), t('Submit'));
+    $this->assertText('URL field is required', "Empty URL submission flagged when URL is required (url_mode: required)", $group);
+    $this->backdropPost(NULL, array('link_test_validate_url[url]' => 'https://backdropcms'), t('Submit'));
+    $this->assertText('The value https://backdropcms provided for Required link with URL validation is not a valid URL', "Invalid URL submission flagged when URL validation is active (validate_url: true)", $group);
+    $this->assertRaw('<label for="edit-configurable-title-and-class-attributes-title">Link "title" attribute ', "Link title attribute is configurable (configurable_title: true)", $group);
+    $this->assertRaw('<label for="edit-configurable-title-and-class-attributes-class">Custom link class ', "Link class attribute is configurable (configurable_class: true)", $group);
+
+    // Test submitted form.
+    $good_submission_data = array(
+      'link_test_defaults[title]' => 'BackdropCMS Demo',
```
Comment (quicksketch) [inline core/modules/link/tests/link.custom.test]:
```suggestion
      'link_test_defaults[title]' => 'Backdrop CMS Demo',
```

File: core/modules/link/tests/link.custom.test [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      'link_test_defaults[url]' => 'https://backdropcms.org/demo',
+      'link_test_title_changes[title]' => 'Upgrade',
+      'link_test_title_changes[url]' => 'https://backdropcms.org/drupal-alternative',
+      'link_test_internal_no_title[url]' => 'about',
+      'link_test_validate_url[title]' => 'Add-ons',
+      'link_test_validate_url[url]' => 'https://backdropcms.org/add-ons',
+      'configurable_title_and_class[title]' => 'Backdrop User Guide',
+      'configurable_title_and_class[url]' => 'https://docs.backdropcms.org/documentation/user-guide',
+      'configurable_title_and_class[attributes][title]' => 'Learn more about using BackdropCMS',
```
Comment (quicksketch) [inline core/modules/link/tests/link.custom.test]:
```suggestion
      'configurable_title_and_class[attributes][title]' => 'Learn more about using Backdrop CMS',
```

File: core/modules/link/tests/link.custom.test [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      'configurable_title_and_class[attributes][class]' => 'help',
+    );
+    $expected_result = '[link_test_defaults] => Array ( [title] => BackdropCMS Demo [url] => https://backdropcms.org/demo ) [link_test_title_changes] => Array ( [title] => Upgrade [url] => https://backdropcms.org/drupal-alternative ) [link_test_internal_no_title] => Array ( [url] => about ) [link_test_validate_url] => Array ( [title] => Add-ons [url] => https://backdropcms.org/add-ons ) [configurable_title_and_class] => Array ( [title] => Backdrop User Guide [url] => https://docs.backdropcms.org/documentation/user-guide [attributes] => Array (  [title] => Learn more about using BackdropCMS  [class] => help ) )';
```
Comment (quicksketch) [inline core/modules/link/tests/link.custom.test]:
```suggestion
    $expected_result = '[link_test_defaults] => Array ( [title] => Backdrop CMS Demo [url] => https://backdropcms.org/demo ) [link_test_title_changes] => Array ( [title] => Upgrade [url] => https://backdropcms.org/drupal-alternative ) [link_test_internal_no_title] => Array ( [url] => about ) [link_test_validate_url] => Array ( [title] => Add-ons [url] => https://backdropcms.org/add-ons ) [configurable_title_and_class] => Array ( [title] => Backdrop User Guide [url] => https://docs.backdropcms.org/documentation/user-guide [attributes] => Array (  [title] => Learn more about using Backdrop CMS  [class] => help ) )';
```

File: core/modules/link/link.module [hunk 5]
```diff
@@ -930,6 +940,54 @@ function link_field_process($element, $form_state, $complete_form) {
...
+/**
+ * Link field element validation.
+ */
+function link_field_element_validate(&$element, &$form_state, $complete_form) {
+  $value = $element['#value'];
+  $is_default_field_value = FALSE;
+
+  // Special validations when the values are not the default and when url has a
+  // value. NB: validation for subelements marked as 'required' is done by core.
+  if (!empty($value['url']) && !$is_default_field_value) {
```
Comment (argiepiano) [inline core/modules/link/link.module]: `$is_defaul_field_value` is always FALSE, and the negation is always TRUE, therefore this is not needed. You can remove the variable `$is_defaul_field_value` completely. This was here before, when the validation checked to see if the value of the element matched the default value. I don't think this is necessary anymore.

Comment (quicksketch) [inline core/modules/link/link.module]:
```suggestion
  // Special validations when the values are not the default and when url has a
  // value. Sub-elements marked as 'required' are validated by core.
  if (!empty($value['url'])) {
```

File: core/modules/link/link.module [hunk 6]
```diff
@@ -813,30 +776,84 @@ function link_theme() {
...
+ *   '#default_value' => array(
+ *     'url' => 'admin/content',
+ *     'title' => 'List of content',
+ *     'attributes' => array( // These are extra elements (see flags below)
+ *       'title' => 'A link to Content', // Prepopulates the "title" attribute.
+ *       'class' => 'additional-css-class',
+ *       'target' => '_blank',
+ *     ),
+ *   ),
+ * @endcode
```
Comment (quicksketch) [inline core/modules/link/link.module:791]:
```suggestion
 * @endcode
 *
 * @since 1.30.0 The "#link_field_options" property added.
```

**Comment by quicksketch** [comment]
Merged into 1.x for 1.30.0. Thank you @argiepiano and @laryn!

**Pull request merged**